python/ray/data/datasource/dbapi2_datasource.py

```diff
@@ -0,0 +1,122 @@
+import math
+from contextlib import contextmanager
+from functools import lru_cache
+from typing import Any, Callable, Generator, Iterable, List, NewType, Optional
+
+import pyarrow as pa
+
+from ray.data._internal.arrow_block import ArrowRow
+from ray.data.block import Block, BlockAccessor, BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask, WriteResult
+
+# A Python DB API2-compliant `Connection` object.
+Connection = Any
+
+
+def cursor_to_block(cursor) -> Block:
+    rows = cursor.fetchall()
+    columns = [column_description[0] for column_description in cursor.description]
+    pylist = [{c: v for c, v in zip(columns, row)} for row in rows]
+    return pa.Table.from_pylist(pylist)
+
+
+class DBAPI2Datasource(Datasource[ArrowRow]):
```

Comment:
Can we rename this to `SQLDatasource` instead? DBAPI2 is the protocol we use to read from SQL database, but we don't read from DBAPI2.